src/monodroid/jni/monodroid-glue.c

```diff
@@ -592,29 +592,84 @@ readdir_r (_WDIR *dirp, struct _wdirent *entry, struct _wdirent **result)
 
 #endif // def WINDOWS
 
-#ifndef RELEASE
+static void
+set_user_executable (const char *path)
+{
+	int r;
+	do
+		r = chmod (path, S_IRUSR | S_IWUSR | S_IXUSR);
+	while (r == -1 && errno == EINTR);
+
+	if (r == -1)
+		log_error (LOG_DEFAULT, "chmod(\"%s\") failed: %s", path, strerror (errno));
+}
 
 static void
-copy_monosgen_to_internal_location(char *to, char *from)
+copy_file_to_internal_location(char *to_dir, char *from_dir, char* file)
 {
-	char *from_libmonoso = path_combine (from, "libmonosgen-2.0.so");
+	char *from_file = path_combine (from_dir, file);
+	char *to_file   = NULL;
+	
+	do {
+		if (!from_file || !file_exists (from_file))
+			break;
 
-	if (!file_exists (from_libmonoso))
-	{
-		free (from_libmonoso);
-		return;
-	}
+		log_warn (LOG_DEFAULT, "Copying file `%s` from external location `%s` to internal location `%s`",
+				file, from_dir, to_dir);
+		
+		to_file = path_combine (to_dir, file);
+		if (!to_file)
+			break;
+		
+		int r = unlink (to_file);
+		if (r < 0 && errno != ENOENT) {
+			log_warn (LOG_DEFAULT, "Unable to delete file `%s`: %s", to_file, strerror (errno));
+			break;
+		}
+		
+		if (file_copy (to_file, from_file) < 0) {
+			log_warn (LOG_DEFAULT, "Copy failed from `%s` to `%s`: %s", from_file, to_file, strerror (errno));
+			break;
+		}
+		
+		set_user_executable (to_file);
+	} while (0);
+	free (from_file);
+	free (to_file);
+}
+
+#ifndef RELEASE
+static void
+copy_native_libraries_to_internal_location (void)
+{
+	int i;
+
+	for (i = 0; i < MAX_OVERRIDES; ++i) {
+		monodroid_dir_t *dir;
+		monodroid_dirent_t b, *e;
 
-	log_warn (LOG_DEFAULT, "Copying sgen from external location %s to internal location %s", from, to);
+		const char *dir_path = path_combine (override_dirs [i], "lib");
+		log_warn (LOG_DEFAULT, "checking directory: `%s`", dir_path);
 
-	char *to_libmonoso = path_combine (to, "libmonosgen-2.0.so");
-	unlink (to_libmonoso);
+		if (dir_path == NULL || !directory_exists (dir_path)) {
+			if (dir_path != NULL)
```

Comment:
No need to check for null. [**free**(3)](https://linux.die.net/man/3/free) will accept the NULL pointer and ignore it:

> The **free()** function.... If ptr is NULL, no operation is performed.